pywbem_mock/_defaultinstanceprovider.py

```diff
@@ -0,0 +1,439 @@
+
+        # The class and instnames are the same except for possible case
+        # sensitivity. If case different, set cprop_name into new instance
+        # to maintain case equality
+        cprop_name = cprop.name
+        if cprop_name != iprop_name:
+            instance.properties[iprop_name].name = cprop_name
+
+        # If property not in instance, add it from class and use default value
+        # from class
+        for cprop_name in target_class.properties:
+            if cprop_name not in instance:
+                default_value = target_class.properties[cprop_name]
+                instance[cprop_name] = default_value
+
+
+class ProviderDispatcher(BaseProvider):
+    """
+    This class dispatches requests destined for the instance provider methods
+    defined in DefaultInstanceProvider  to the either the default instance
+    provder method or the provider defined for processing the defined class in
+    the defined namespace.
+    """
+    def __init__(self, conn, default_instance_provider):
+        """
+        """
+        # TODO: This not correct way to make cimrepository available to
+        # multiple subclasses
+        super(ProviderDispatcher, self).__init__(
+            conn=None, cimrepository=None)
+        self.default_instance_provider = default_instance_provider
+
+        self.cimrepository = conn.cimrepository
+        self.provider_registry = conn.provider_registry
+
+    def get_registered_provider(self, namespace, provider_type, classname):
+        """
+        If there is a provider registered for this namespace and classname
+        return that object (the instance of the DefaultInstanceProvider
+        subclass).
+
+        If no provider is registered, return None
+        """
+        if self.provider_registry:
+            if classname not in self.provider_registry:
+                return None
+            if self.provider_registry[classname]['providertype'] != provider_type:
+                return None
+            if not self.provider_registry[classname]['namespaces']:
+                return None
+            if namespace in self.provider_registry[classname]['namespaces']:
+                return self.provider_registry[classname]['provider']
+        return None
+
+    def CreateInstance(self, namespace, NewInstance):
+        """
+        Dispatcher for CreateInstance.
+
+        This method validates input parameters against the repository and
+        provider required types and routes the method either to the default
+        provider or to a provider registered for the namespace and class in
+        InstanceName
+        """
+        # Validate input parameters
+
+        self.validate_namespace(namespace)
+
+        if not isinstance(NewInstance, CIMInstance):
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("NewInstance parameter is not a valid CIMInstance. "
+                        "Rcvd type={0}", type(NewInstance)))
```

Comment:
In CIM-XML, this would be detected before the request is sent to the server because the invalid type cannot be represented in CIM-XML. I therefore believe this should result in a TypeError (but I did not check what we do for CIM-XML).

Reply:
Agree that in CIM-XML this would be detected. So the dispatcher really does not even need to test for it because the client/imethodcall/us also checks for it.  However, maybe it should be in the DefaultInstanceProvider since the user provider could make a mess and hand it back to the default provider,

**DISCUSSION:**

Reply:
Agree change to TypeError